File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter05/DependencyProperties/DependencyProperties/GradientButton.cpp
#include "pch.h"
#include "GradientButton.h"

using namespace DependencyProperties;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Controls::Primitives;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Navigation;

DependencyProperty^ GradientButton::color1Property = 
    DependencyProperty::Register("Color1", 
                                 TypeName(Color::typeid), 
                                 TypeName(GradientButton::typeid),
                                 ref new PropertyMetadata(Colors::White,
                                 ref new PropertyChangedCallback(GradientButton::OnColorChanged)));

DependencyProperty^ GradientButton::color2Property = 
    DependencyProperty::Register("Color2", 
                                 TypeName(Color::typeid), 
                                 TypeName(GradientButton::typeid), 
                                 ref new PropertyMetadata(Colors::Black,
                                 ref new PropertyChangedCallback(GradientButton::OnColorChanged)));

GradientButton::GradientButton()
{
    gradientStop1 = ref new GradientStop();
    gradientStop1->Offset = 0;
    gradientStop1->Color = this->Color1;

    gradientStop2 = ref new GradientStop();
    gradientStop2->Offset = 1;
    gradientStop2->Color = this->Color2;

    LinearGradientBrush^ brush = ref new LinearGradientBrush();
    brush->GradientStops->Append(gradientStop1);
    brush->GradientStops->Append(gradientStop2);

    this->Foreground = brush;
}

DependencyProperty^ GradientButton::Color1Property::get()
{
    return color1Property;
}

DependencyProperty^ GradientButton::Color2Property::get()
{
    return color2Property;
}

Color GradientButton::Color1::get()
{
    return Color(GetValue(Color1Property));
}

void GradientButton::Color1::set(Color value)
{
    SetValue(Color1Property, value);
}

Color GradientButton::Color2::get()
{
    return Color(GetValue(Color2Property));
}

void GradientButton::Color2::set(Color value)
{
    SetValue(Color2Property, value);
}

void GradientButton::OnColorChanged(DependencyObject^ sender, DependencyPropertyChangedEventArgs^ args)
{
    dynamic_cast<GradientButton^>(sender)->OnColorChangedInstance(args);
}

void GradientButton::OnColorChangedInstance(DependencyPropertyChangedEventArgs^ args)
{
    if (args->Property == GradientButton::Color1Property)
        gradientStop1->Color = Color(args->NewValue);

    if (args->Property == GradientButton::Color2Property)
        gradientStop2->Color = Color(args->NewValue);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter05/DependencyProperties/DependencyProperties/GradientButton.cs
using Windows.UI;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;

namespace DependencyProperties
{
    public class GradientButton : Button
    {
        GradientStop gradientStop1, gradientStop2;

        static GradientButton()
        {
            Color1Property =
                DependencyProperty.Register("Color1",
                    typeof(Color),
                    typeof(GradientButton),
                    new PropertyMetadata(Colors.White, OnColorChanged));

            Color2Property =
                DependencyProperty.Register("Color2",
                    typeof(Color),
                    typeof(GradientButton),
                    new PropertyMetadata(Colors.Black, OnColorChanged));
        }

        public static DependencyProperty Color1Property { private set; get; }

        public static DependencyProperty Color2Property { private set; get; }

        public GradientButton()
        {
            gradientStop1 = new GradientStop
            {
                Offset = 0,
                Color = this.Color1
            };

            gradientStop2 = new GradientStop
            {
                Offset = 1,
                Color = this.Color2
            };

            LinearGradientBrush brush = new LinearGradientBrush();
            brush.GradientStops.Add(gradientStop1);
            brush.GradientStops.Add(gradientStop2);

            this.Foreground = brush;
        }

        public Color Color1
        {
            set { SetValue(Color1Property, value); }
            get { return (Color)GetValue(Color1Property); }
        }

        public Color Color2
        {
            set { SetValue(Color2Property, value); }
            get { return (Color)GetValue(Color2Property); }
        }

        static void OnColorChanged(DependencyObject obj,
                                   DependencyPropertyChangedEventArgs args)
        {
            (obj as GradientButton).OnColorChanged(args);
        }

        void OnColorChanged(DependencyPropertyChangedEventArgs args)
        {
            if (args.Property == Color1Property)
                gradientStop1.Color = (Color)args.NewValue;

            gradientStop1.Color = this.Color1;

            if (args.Property == Color2Property)
                gradientStop2.Color = (Color)args.NewValue;
        }
    }
}